File: src/libsegmenter/windows/hann50.py
# Copyright (c) 2025 Niels de Koeijer, Martin Bo Møller
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
# the Software, and to permit persons to whom the Software is furnished to do so,
# subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
# FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

import numpy as np

from libsegmenter.Window import Window


def hann50(segment_size: int) -> Window:
    """
    Generates a Hann window of the given size.

    Parameters:
    -----------
    window_size : int
        The size of the window.

    Returns:
    --------
    Window
        A Window describing the Hann window values.
    """

    assert (segment_size % 2 == 0, f"segment_size must be even, got {segment_size}")

    M = float(segment_size)
    indices = np.arange(segment_size)
    window = 0.5 * (1.0 - np.cos(2.0 * np.pi * indices / M))

    return Window(segment_size // 2, window, window)

File: src/libsegmenter/windows/hann50.m
function [window, hopSize] = hann50(segmentSize)
    % Generates a Hann window of the given size with 50% overlap.
    %
    % Args:
    %   segmentSize (double): Size of the window to be created.
    %
    % Returns:
    %   A Hann window with 50% overlap
    if length(segmentSize) ~= 1
        error(['The segment size must be a scalar. Received size(segmentSize) = ' int2str(size(segmentSize)) ]);
    end
    if mod(segmentSize,2)
        error(['Segment size must be even, got ' int2str(segmentSize) ]);
    end
    indices = (0:segmentSize-1)';
    window = hannWindow(segmentSize, indices);
    hopSize = segmentSize/2;
end
